internal/generator/backend/template/zkpschemes/plonk/plonk.setup.go.tmpl

```diff
@@ -75,143 +58,266 @@ type VerifyingKey struct {
+
+	// step 3: build the permutation and build the polynomials S1, S2, S3 to encode the permutation.
+	// Note: at this stage, the permutation takes in account the placeholders
+	nbVariables := spr.NbInternalVariables + len(spr.Public) + len(spr.Secret)
+	buildPermutation(spr, &pk.trace, nbVariables)
+	s := computePermutationPolynomials(&pk.trace, &pk.Domain[0])
+	pk.trace.S1 = s[0]
+	pk.trace.S2 = s[1]
+	pk.trace.S3 = s[2]
+
+	// step 4: commit to s1, s2, s3, ql, qr, qm, qo, and (the incomplete version of) qk.
+	// All the above polynomials are expressed in canonical basis afterwards. This is why
+	// we save lqk before, because the prover needs to complete it in Lagrange form, and
+	// then express it on the Lagrange coset basis.
+	pk.lQk = pk.trace.Qk.Clone() // it will be completed by the prover, and the evaluated on the coset
+	err := commitTrace(&pk.trace, &pk)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// step 5: evaluate ql, qr, qm, qo, s1, s2, s3 on LagrangeCoset (NOT qk)
+	// we clone them, because the canonical versions are going to be used in
+	// the opening proof
+	pk.computeLagrangeCosetPolys()
+
+	return &pk, &vk, nil
+}
+
+// computeLagrangeCosetPolys computes each polynomial except qk in Lagrange coset
+// basis. Qk will be evaluated in Lagrange coset basis once it is completed by the prover.
+func (pk *ProvingKey) computeLagrangeCosetPolys() {
+	pk.lcQl = pk.trace.Ql.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcQr = pk.trace.Qr.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcQm = pk.trace.Qm.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcQo = pk.trace.Qo.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcS1 = pk.trace.S1.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcS2 = pk.trace.S2.Clone().ToLagrangeCoset(&pk.Domain[1])
+	pk.lcS3 = pk.trace.S3.Clone().ToLagrangeCoset(&pk.Domain[1])
+}
+
+// NbPublicWitness returns the expected public witness size (number of field elements)
+func (vk *VerifyingKey) NbPublicWitness() int {
+	return int(vk.NbPublicVariables)
+}
+
+// VerifyingKey returns pk.Vk
+func (pk *ProvingKey) VerifyingKey() interface{} {
+	return pk.Vk
+}
+
+// InitKZG inits pk.Vk.KZG using pk.Domain[0] cardinality and provided SRS
+//
+// This should be used after deserializing a ProvingKey
+// as pk.Vk.KZG is NOT serialized
+func (pk *ProvingKey) InitKZG(srs kzgg.SRS) error {
+	return pk.Vk.InitKZG(srs)
+}
+
+// InitKZG inits vk.KZG using provided SRS
+//
+// This should be used after deserializing a VerifyingKey
+// as vk.KZG is NOT serialized
+//
+// Note that this instantiate a new FFT domain using vk.Size
+func (vk *VerifyingKey) InitKZG(srs kzgg.SRS) error {
+	_srs := srs.(*kzg.SRS)
+
+	if len(_srs.G1) < int(vk.Size) {
+		return errors.New("kzg srs is too small")
+	}
+	vk.KZGSRS = _srs
+
+	return nil
+}
+
+// BuildTrace fills the constatn columns ql, qr, qm, qo, qk from the sparser1cs.
+// Size is the size of the system that is nb_constraints+nb_public_variables
+func BuildTrace(spr *cs.SparseR1CS, pt *PlonkTrace) {
```

Comment:
needs to be exposed? If so; doesn't it make more sense to be: 
`plonk.NewTrace(ccs *cs.SparseR1CS) (*PlonkTrace, error)` ?

Reply:
I exposed it so it can be used by the wizard iop or any tool that needs a complete trace of a plonk circuit